File: static/pitchbend_overlay.js
/**
 * Helper functions for the pitch bend overlay displayed in the clip editor.
 *
 * ``BASE_NOTE`` and ``SEMI_UNIT`` mirror the values used by Ableton Live to
 * convert between pitch bend values and semitone offsets.
 */
export const BASE_NOTE = 36;
export const SEMI_UNIT = 170.6458282470703;

export function noteNumberToPitchbend(n) {
  return (n - BASE_NOTE) * SEMI_UNIT;
}

export function computeOverlayNotes(sequence, selectedRow, ticksPerBeat) {
  // Generate an array of note objects visualising pitch bend edits.
  const overlay = [];
  if (selectedRow === null || selectedRow === undefined) return overlay;
  if (!sequence || !ticksPerBeat) return overlay;
  sequence.forEach((ev, idx) => {
    if (ev.n !== selectedRow) return;
    // Treat missing or zero-length PitchBend as value 0 (C2)
    const pb = ev.a && ev.a.PitchBend;
    const value = (pb && pb.length) ? pb[0].value : 0;
    const semis = Math.round(value / SEMI_UNIT);
    const viz = BASE_NOTE + semis;
    if (viz < 0 || viz > 127) return;
    overlay.push({
      index: idx,
      noteNumber: viz,
      startTime: ev.t / ticksPerBeat,
      duration: ev.g / ticksPerBeat
    });
  });
  return overlay;
}

File: core/pitchbend_overlay.py
from typing import List, Dict, Any

BASE_NOTE = 36
SEMI_UNIT = 170.6458282470703


def compute_overlay_notes(
    notes: List[Dict[str, Any]],
    selected_row: int,
    base_note: int = BASE_NOTE,
    semi_unit: float = SEMI_UNIT,
) -> List[Dict[str, float]]:
    """Return overlay notes for the given row using pitch-bend automation."""
    overlay: List[Dict[str, float]] = []
    if selected_row is None:
        return overlay
    for note in notes:
        if int(note.get("noteNumber", -1)) != int(selected_row):
            continue
        autom = note.get("automations") or {}
        pb = None
        if isinstance(autom, dict):
            pb = autom.get("PitchBend")
        if not pb:
            continue
        first = pb[0] if pb else None
        if not first:
            continue
        value = first.get("value")
        if value is None:
            continue
        try:
            semis = round(float(value) / semi_unit)
        except Exception:
            continue
        viz = base_note + semis
        if 0 <= viz <= 127:
            overlay.append({
                "noteNumber": viz,
                "startTime": float(note.get("startTime", 0.0)),
                "duration": float(note.get("duration", 0.0)),
            })
    return overlay
